### PR TITLE
Add analytics event types

### PR DESCRIPTION
**Pull request by josemiguel-alvarez**
### What are you trying to accomplish?

Add analytics event types. The types must match the definitions of https://github.com/Shopify/ui-extensions/blob/main/packages/web-pixels-extension/src/types/PixelEvents/index.ts

### Before you deploy

- [ ] I have added tests to support my implementation
- [ ] I have read and agree with the contributing documentation [readme](https://github.com/shopify/mobile-checkout-sdk-ios/blob/main/.github/CONTRIBUTING.md)
- [ ] I have read and agree with the code of conduct documentation [readme](https://github.com/shopify/mobile-checkout-sdk-ios/blob/main/.github/CODE_OF_CONDUCT.md)
- [ ] I have updated any documentation related to these changes.
- [ ] I have updated the [README](https://github.com/shopify/mobile-checkout-sdk-ios) (if applicable).

### CODE REVIEW

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 1]
```diff
@@ -0,0 +1,4594 @@
...
+    }
+}
+
+/// The name of the customer event
+enum PixelEventsSearchSubmittedName: String, Codable {
+    case searchSubmitted = "search_submitted"
+}
+
+// MARK: - Browser
+struct Browser: Codable {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3455]: I think this is unrelated to events and can be dropped, along with the types it references `BrowserCookie`, `SendBeacon` etc)

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 2]
```diff
@@ -0,0 +1,4594 @@
...
+        checkoutStarted: PixelEventsCheckoutStarted?? = nil,
+        collectionViewed: PixelEventsCollectionViewed?? = nil,
+        pageViewed: PixelEventsPageViewed?? = nil,
+        paymentInfoSubmitted: PixelEventsPaymentInfoSubmitted?? = nil,
+        productAddedToCart: PixelEventsProductAddedToCart?? = nil,
+        productRemovedFromCart: PixelEventsProductRemovedFromCart?? = nil,
+        productVariantViewed: PixelEventsProductVariantViewed?? = nil,
+        productViewed: PixelEventsProductViewed?? = nil,
+        searchSubmitted: PixelEventsSearchSubmitted?? = nil
+    ) -> PixelEvents {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:206]: I'm not sure we'll need this function? It looks like it returns one of each pixel event

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 3]
```diff
@@ -0,0 +1,4594 @@
...
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// The `cart_viewed` event logs an instance where a customer visited the cart
+/// page
+// MARK: - PixelEventsCartViewed
+struct PixelEventsCartViewed: Codable {
+    /// The client-side ID of the customer, provided by Shopify
+    let clientID: String?
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:239]: I believe we should be dropping `clientID`, lets double check

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 4]
```diff
@@ -0,0 +1,4594 @@
...
+    /// The ID of the customer event
+    let id: String?
+    /// The name of the customer event
+    let name: PixelEventsCartViewedName?
+    /// The timestamp of when the customer event occurred, in [ISO
+    /// 8601](https://en.wikipedia.org/wiki/ISO_8601) format
+    let timestamp: String?
+
+    enum CodingKeys: String, CodingKey {
+        case clientID = "clientId"
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:251]: we can get rid of this mapping if we remove clientId

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 5]
```diff
@@ -0,0 +1,4594 @@
...
+        try self.init(data: data)
+    }
+
+    init(fromURL url: URL) throws {
+        try self.init(data: try Data(contentsOf: url))
+    }
+
+    func with(
+        clientID: String?? = nil,
+        context: Context?? = nil,
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:276]: I think events each have a context, id, name, timestamp with the same type.

Can we introduce an interface or supertype?

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 6]
```diff
@@ -0,0 +1,4594 @@
...
+//   let timestamp = try Timestamp(json)
+//   let transaction = try Transaction(json)
+//   let webPixelsDocument = try WebPixelsDocument(json)
+//   let webPixelsNavigator = try WebPixelsNavigator(json)
+//   let webPixelsWindow = try WebPixelsWindow(json)
+
+import Foundation
+
+// MARK: - PixelEvents
+struct PixelEvents: Codable {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:99]: Is it worth adding a test around deserializing JSON into one or two events.

Maybe also with missing fields, to double check optionality.

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 7]
```diff
@@ -0,0 +1,3579 @@
...
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+// MARK: - PixelEventsPageViewedData
+struct PixelEventsPageViewedData: Codable {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:2519]: This one isn't typed very well in web pixels manager. Not sure if there's much we can do about it?

File: Sources/ShopifyCheckoutKit/AnalyticsEventDecoder.swift [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+        case name
+        case event
+    }
+    
+    struct AnalyticsEvent: Decodable {
+        let name: String
+        let event: AnalyticsEventBody
+    }
+    
+    struct AnalyticsEventBody: Decodable {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEventDecoder.swift:37]: you should be able to `swiftlint --fix` these

You might need to `brew install` first
https://github.com/realm/SwiftLint

File: Samples/MobileBuyIntegration/MobileBuyIntegration/CartViewController.swift [hunk 9]
```diff
@@ -151,7 +151,7 @@ class CartViewController: UIViewController, UITableViewDelegate, UITableViewData
 	}
 }
 
-extension CartViewController: CheckoutDelegate {
+extension CartViewController: CheckoutDelegate {
```
Comment (kiftio) [inline Samples/MobileBuyIntegration/MobileBuyIntegration/CartViewController.swift:154]: lets get rid of the trailing whitespace here

File: Sources/ShopifyCheckoutKit/AnalyticsEventDecoder.swift [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            default:
+                customData = nil
+                data = nil
+                break
+            }
+
+        }
+    }
+    
+    func decode(from container: KeyedDecodingContainer<CheckoutBridge.WebEvent.CodingKeys>, using decoder: Decoder) throws -> Decodable?  {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEventDecoder.swift:83]: Wondering if returning a Decodable is a bit too generic as it doesn't narrow the possible values down to only analytics events

If it was a specific super type, we could also move common attributes there, id, name, timestamp etc, and allow access to those properties without knowing the specific sub-type.

Interested to see what others think. 

Think it'd also be good to write some tests doing what we expect a client might do to see if we hit any issues

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 11]
```diff
@@ -0,0 +1,3579 @@
...
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+// This file was generated from JSON Schema using quicktype, do not modify it directly.
+// To parse the JSON, add this file to your project and do:
+//
+//   let pixelEventsCartViewedData = try PixelEventsCartViewedData(json)
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:27]: Do you think we should remove the long comment?

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 12]
```diff
@@ -0,0 +1,3579 @@
...
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// A free-form object representing data specific to this event provided by
+/// Shopify. Refer to [standard events](#standard-events) for details on the
+/// payload available to each event
+// MARK: - PurpleData
+struct PurpleData: Codable {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3366]: what is purple data?

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift [hunk 13]
```diff
@@ -0,0 +1,3579 @@
...
+/// percentage off a specified item.
+// MARK: - PricingPercentageValue
+struct PricingPercentageValue: Codable {
+    /// The percentage value of the object.
+    let percentage: Double?
+}
+
+// MARK: PricingPercentageValue convenience initializers and mutators
+
+extension PricingPercentageValue {
```
Comment (kiftio) [inline Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3526]: Have we also covered DOM events?

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift (hunk 12 above)
Comment (josemiguel-alvarez) [reply Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3366]: Not sure why it was added, I cannot find it in the original code. I'll remove it.

Comment (josemiguel-alvarez) [reply Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3366]: It looks like that it refers to this:

```
/**
 * A free-form object representing data specific to this event provided by
 * Shopify. Refer to [standard events](#standard-events) for details on the
 * payload available to each event
 */
export type Data = Record<string, unknown>;
```

It is not being used, so I'll remove it.

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift (hunk 13 above)
Comment (josemiguel-alvarez) [reply Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3526]: Are they already included in the Web Pixels Manager?

Comment (kiftio) [reply Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3526]: Not yet, but think they will be very soon

File: Samples/MobileBuyIntegration/MobileBuyIntegration/CartViewController.swift [hunk 14]
```diff
@@ -183,6 +183,10 @@ extension CartViewController: CheckoutDelegate {
 		}
 	}
 
+    func checkoutDidEventAnalyticsEvent(decodable: Decodable) {
```
Comment (cianBuckley) [inline Samples/MobileBuyIntegration/MobileBuyIntegration/CartViewController.swift:186]: This should be called `checkoutDidEmitWebPixel`

File: Sources/ShopifyCheckoutKit/AnalyticsEvents.swift (hunk 13 above)
Comment (kiftio) [reply Sources/ShopifyCheckoutKit/AnalyticsEvents.swift:3415]: sure, we can come back to it

File: Tests/ShopifyCheckoutKitTests/CheckoutBridgeTests.swift [hunk 15]
```diff
@@ -113,6 +113,21 @@ class CheckoutBridgeTests: XCTestCase {
...
...23\\\",\\\"name\\\":\\\"cart_viewed\\\",\\\"type\\\":\\\"standard\\\",\\\"timestamp\\\":\\\"2024-01-02T14:14:59.699Z\\\",\\\"data\\\":{\\\"cart\\\":{\\\"id\\\":\\\"123\\\"}},\\\"customData\\\":null,\\\"context\\\":{\\\"document\\\":{\\\"characterSet\\\":\\\"\\\",\\\"location\\\":{\\\"hash\\\":\\\"\\\",\\\"host\\\":\\\"\\\",\\\"hostname\\\":\\\"\\\",\\\"href\\\":\\\"\\\",\\\"origin\\\":\\\"\\\",\\\"pathname\\\":\\\"\\\",\\\"port\\\":\\\"\\\",\\\"protocol\\\":\\\"\\\",\\\"search\\\":\\\"\\\"},\\\"referrer\\\":\\\"\\\",\\\"title\\\":\\\"\\\"},\\\"navigator\\\":{\\\"cookieEnabled\\\":false,\\\"language\\\":\\\"\\\",\\\"languages\\\":[],\\\"userAgent\\\":\\\"\\\"},\\\"window\\\":{\\\"innerHeight\\\":0,\\\"innerWidth\\\":0,\\\"location\\\":{\\\"hash\\\":\\\"\\\",\\\"host\\\":\\\"\\\",\\\"hostname\\\":\\\"\\\",\\\"href\\\":\\\"\\\",\\\"origin\\\":\\\"\\\",\\\"pathname\\\":\\\"\\\",\\\"port\\\":\\\"\\\",\\\"protocol\\\":\\\"\\\",\\\"search\\\":\\\"\\\"},\\\"origin\\\":\\\"\\\",\\\"outerHeight\\\":0,\\\"outerWidth\\\":0,\\\"pageXOffset\\\":0,\\\"pageYOffset\\\":0,\\\"screen\\\":{\\\"height\\\":0,\\\"width\\\":0},\\\"screenX\\\":0,\\\"screenY\\\":0,\\\"scrollX\\\":0,\\\"scrollY\\\":0}}}}\"}"
```
Comment (cianBuckley) [inline Tests/ShopifyCheckoutKitTests/CheckoutBridgeTests.swift:120]: I think you can continue with the multiline pattern you had on name:analytics and remove all those escapes? If not then you could pull the `body` out and embed it as a variable.
